File: solutions/csharp/proverb/1/Proverb.cs
using System;

public static class Proverb
{
    public static string[] Recite(string[] subjects)
    {
        int proverbLength = subjects.Length;
        var proverbLines = new string[proverbLength];

        for (int i = 0; i < proverbLength; i++)
        {
            if (i == proverbLength - 1)
            {
                proverbLines[i] = $"And all for the want of a {subjects[0]}.";
            }
            else
            {
                proverbLines[i] = $"For want of a {subjects[i]} the {subjects[i + 1]} was lost.";
            }
        }

        return proverbLines;
    }
}

File: solutions/fsharp/proverb/1/Proverb.fs
module Proverb

let recite (input: string list): string list = 
    match input with
    | [] ->
        []
    | [one] ->
        [$"And all for the want of a {one}."]
    | _ ->
        let shiftedInput = List.permute (fun x -> (x + List.length input - 1) % List.length input) input
        let shiftedOutput = List.map2 (fun x y -> $"For want of a {x} the {y} was lost.") input shiftedInput |> List.take (List.length input - 1)
        List.append shiftedOutput [$"And all for the want of a {List.head input}."]